indexing-service/src/main/java/org/apache/druid/indexing/appenderator/ActionBasedPublishedSegmentRetriever.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.druid.indexing.common.actions.RetrieveUsedSegmentsAction;
+import org.apache.druid.indexing.common.actions.TaskActionClient;
+import org.apache.druid.indexing.overlord.Segments;
+import org.apache.druid.java.util.common.JodaUtils;
+import org.apache.druid.segment.realtime.appenderator.PublishedSegmentRetriever;
+import org.apache.druid.timeline.DataSegment;
+import org.apache.druid.timeline.SegmentId;
+import org.joda.time.Interval;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ActionBasedPublishedSegmentRetriever implements PublishedSegmentRetriever
+{
+  private final TaskActionClient taskActionClient;
+
+  public ActionBasedPublishedSegmentRetriever(TaskActionClient taskActionClient)
+  {
+    this.taskActionClient = taskActionClient;
+  }
+
+  @Override
+  public Set<DataSegment> findPublishedSegments(Set<SegmentId> segmentIds) throws IOException
+  {
+    // Validate that all segments belong to the same datasource
+    final String dataSource = segmentIds.iterator().next().getDataSource();
+    final Set<SegmentId> segmentIdsToFind = new HashSet<>(segmentIds);
+    for (SegmentId segmentId : segmentIds) {
+      if (!segmentId.getDataSource().equals(dataSource)) {
+        throw InvalidInput.exception(
+            "Published segment IDs to find cannot belong to multiple datasources[%s, %s].",
+            dataSource, segmentId.getDataSource()
+        );
+      }
+    }
+
+    final Set<DataSegment> publishedSegments = new HashSet<>();
+
+    // Search for the required segments in the "used" set
+    final List<Interval> usedSearchIntervals = JodaUtils.condenseIntervals(
+        Iterables.transform(segmentIdsToFind, SegmentId::getInterval)
+    );
+    final Collection<DataSegment> foundUsedSegments = taskActionClient.submit(
+        new RetrieveUsedSegmentsAction(dataSource, null, usedSearchIntervals, Segments.INCLUDING_OVERSHADOWED)
+    );
+    for (DataSegment segment : foundUsedSegments) {
+      if (segmentIdsToFind.contains(segment.getId())) {
+        publishedSegments.add(segment);
+        segmentIdsToFind.remove(segment.getId());
+      }
+    }
+
+    if (segmentIdsToFind.isEmpty()) {
+      return publishedSegments;
+    }
+
+    // Search for the remaining segments in the "unused" set
+    final Set<String> versions = segmentIdsToFind.stream().map(SegmentId::getVersion).collect(Collectors.toSet());
+    final List<Interval> unusedSearchIntervals = JodaUtils.condenseIntervals(
+        Iterables.transform(segmentIdsToFind, SegmentId::getInterval)
+    );
+    for (Interval searchInterval : unusedSearchIntervals) {
+      final Collection<DataSegment> foundUnusedSegments = taskActionClient.submit(
+          new RetrieveUnusedSegmentsAction(dataSource, searchInterval, new ArrayList<>(versions), null, null)
```

Comment:
This action has been known to be slow for clusters with a large number of unused segments.
It might be better to not condense the intervals and fetch segments for each matching interval individually.
Perhaps the query to fetch ids in https://github.com/apache/druid/pull/16623 can be modified to fetch the payloads instead.

Reply:
Is there a reason we need to match DataSegment objects.
Why is it insufficient to check that the segments exist in the metadata store?

Reply:
We do need the load spec of the segments too to check if they are disjoint. Let me check what we can do about using the sub-optimal APIs.